src/swarm/README_protocol_neo.md

```diff
@@ -0,0 +1,146 @@
+
+The request protocol works as follows:
+* The first 64 bits of all message payloads contains the id of a request.
+* Using this id, the client/node can look up the request to which a message is
+  directed, and notify it that a message has arrived.
+* If a message is received for a request id that does not exist, the message is
+  simply discarded, with no effect.
+
+Apart from the request id, the message payload is opaque data, at this level,
+specified by the protocol for the specific request type that is sending the
+message.
+
+# Connection and Request Handling
+
+Connection and request handling vary significantly between the client and node,
+however there are many similarities. These are discussed in this section.
+
+## Connections
+
+The `ConnectionBase` class (`swarm.neo.connection.ConnectionBase`) provides the
+basic functionality of a full-duplex connection. It contains:
+* Two fibers, one handling sending over the connection and one handling
+  receiving. The send and receive fibers both run a loop, internally, waiting
+  for a message to be provided for sending or to arrive from the connection,
+  handling the operation, then returning to waiting.
+* Message dispatching logic that reads the request id from the message payload
+  (the first 64 bits), looks up the appropriate request, and passes the
+  remainder of the message payload to it.
+* Logic for registering the connection socket with epoll. (A `ConnectionBase`
+  object is an `ISelectClient`. See
+  `ocean.io.select.client.model.ISelectClient`.)
+
+### The Send Queue
+
+The send fiber owned by each connection has an integrated queue of requests that
+are waiting to send something. When a request wishes to send, it must register
+itself with the appropriate connection; its message will be sent down the
+connection after all other pending writes have completed.
+
+## Requests on Connections
+
+A request can require access to any number of connections to send/receive over.
+An instance of the `RequestOnConnBase` class
+(`swarm.neo.connection.RequestOnConnBase`) denotes a unique combination of a
+_request_ operating over a specific _connection_ (a request-on-connection),
+providing the core functionality required for a request to transmit messages
+over the connection . Requests on connections are commonly referred to as
+"request-on-conns" or as "RoCs".
+
+A request-on-conn has the following important components:
+* A fiber that manages asynchronous I/O operations for the request on the
+  connection.
+* An event dispatcher (the nested class `EventDispatcher`) instance that
+  provides the following functionality:
+  - A method to suspend the RoC fiber and wait for one of a set of specified
+    events. Events that can be waited on are:
+    1. Receiving a payload for this request.
+    2. Sending a payload.
+    3. The fiber being resumed after it was yielded and an epoll event-loop
+       cycle occurred.
+    4. The fiber being resumed with a non-negative code.
```

Comment:
It might be worth mentioning that the fiber is here _explicitly_ resumed.

Reply:
Hmm, what do you mean by "explicitly"? Is it possible for a fiber to be non-explicitly resumed?

Reply:
By explicitly calling `resumeFiber` method. AFAIR, all other cases resume fiber with the negative code, right? I think it's useful to say that the non-negative codes are reserved for explicit `resumeFiber` calls.

Reply:
Ping :-)

Reply:
Updated.

Reply:
Thanks!